sled-agent/src/bootstrap/server.rs

```diff
@@ -154,27 +231,110 @@ impl Inner {
     }
 }
 
-async fn serve_single_request(
+async fn serve_request_before_quorum_initialization(
     stream: TcpStream,
     sp: Option<SpHandle>,
-    bootstrap_agent: Arc<Agent>,
+    bootstrap_agent: &Agent,
+    tx_share: mpsc::Sender<Option<ShareDistribution>>,
     log: &Logger,
 ) -> Result<(), String> {
+    // Establish sprockets session (if we have an SP).
+    let mut stream = crate::sp::maybe_wrap_stream(
+        stream,
+        &sp,
+        SprocketsRole::Server,
+        None, // We don't have our trust quorum members yet
+        log,
+    )
+    .await
+    .map_err(|err| format!("Failed to establish sprockets session: {err}"))?;
+
+    let response = match read_request(&mut stream).await? {
+        Request::SledAgentRequest(request) => {
+            match bootstrap_agent.request_agent(&*request).await {
+                Ok(response) => {
+                    // Bootstrap agent successfully initialized our sled; tell
+                    // `wait_for_sled_initialization()` what our trust quorum
+                    // share is. It doesn't return until we tell it our share
+                    // (unless something has gone wrong with the socket), so
+                    // it's fine to unwrap this send.
```

Comment:
In conjunction with my comment above: While this appears to be correct, I'm finding it hard to reason about. Can the following scenario occur?

Let's say RSS doesn't get a response in time, and initiates a new connection before the call to `tx_share.send()` occurs. A new task gets spawned that calls `serve_request_before_quorum_initialization` from `wait_for_sled_initialization` with a clone of `tx_share`. The  sled agent returns a response and this call to `tx_share.send` completes which causes the loop to terminate in `wait_for_sled_initialization` and the receiver, `rx_share` to be dropped. The second request now returns from the sled agent and `tx_share.send` is called again, but with a dropped receiver. This will cause a panic on `unwrap`. 

While I'm not sure this scenario is possible in our current code base, it's very hard to tell this from local reasoning.

Reply:
Nice catch, I think this is possible (albeit unlikely). I'll change the `.unwrap()` to just ignoring the result, since at the point of this send we don't care if our caller has gone away.

Reply:
Changed in ef6a0007

Reply:
Works for me!